src/openvic-simulation/history/ProvinceHistory.cpp

```diff
@@ -0,0 +1,260 @@
+        if (entry.second.size() == 0) {
+            Logger::error("Attempted to lock province histories - province ", entry.first->get_identifier(), " has no history entries!");
+        }
+    }
+    Logger::info("Locked province history registry after registering ", province_histories.size(), " items");
+    locked = true;
+}
+
+bool ProvinceHistoryManager::is_locked() const {
+    return locked;
+}
+
+ProvinceHistory const* ProvinceHistoryManager::get_province_history(Province const* province, Date entry) const {
+    Date closest_entry;
+    auto province_registry = province_histories.find(province);
+
+    if (province_registry == province_histories.end()) {
+        Logger::error("Attempted to access history of undefined province ", province->get_identifier());
+        return nullptr;
+    }
+
+    for (const auto& current : province_registry->second) {
+        if (current.first == entry) return &current.second;
+        if (current.first > entry) continue;
+        if (current.first > closest_entry && current.first < entry) closest_entry = current.first;
+    }
+
+    auto entry_registry = province_registry->second.find(closest_entry);
+    if (entry_registry != province_registry->second.end()) {
+        return &entry_registry->second;
+    }
+	/* warned about lack of entries earlier, return nullptr */
+	return nullptr;
+}
+
+inline ProvinceHistory const* ProvinceHistoryManager::get_province_history(Province const* province, Bookmark const* bookmark) const {
+    return get_province_history(province, bookmark->get_date());
+}
+
+inline bool ProvinceHistoryManager::_load_province_history_entry(GameManager& game_manager, std::string_view province, Date const& date, ast::NodeCPtr root) {
+    Country const* owner = nullptr;
+    Country const* controller = nullptr;
+    std::vector<Country const*> cores;
+    Good const* rgo;
+    uint8_t life_rating = 0;
+    TerrainType const* terrain_type;
+    std::map<Building const*, uint8_t> buildings;
+    std::map<Ideology const*, uint8_t> party_loyalties;
+    
+    bool updated_cores = false;
+    bool updated_buildings = false;
+    bool updated_loyalties = false;
+
+    bool ret = expect_dictionary_keys_and_default(
+        [&game_manager, &buildings, &updated_buildings](std::string_view key, ast::NodeCPtr value) -> bool {
+            // used for province buildings like forts or railroads
+            if (game_manager.get_economy_manager().get_building_manager().has_building_identifier(key)) {
+                Building const* building;
+                uint8_t level;
+
+                bool ret = game_manager.get_economy_manager().get_building_manager().expect_building_str(assign_variable_callback_pointer(building))(key);
+                ret &= expect_uint(assign_variable_callback(level))(value);
+
+                buildings.emplace(building, level);
+                updated_buildings = true;
+                return ret;
+            }
+            return true;
```

Comment:
```suggestion
            return key_value_invalid_callback(key, value);
```
Currently invalid keys are ignored - a simple solution would be to return `false` instead, but that wouldn't print an error message. Using `key_value_invalid_callback` can save you the trouble of writing that error message out yourself.